kostyor/upgrades/driver.py

```diff
@@ -0,0 +1,85 @@
+import abc
+import six
+
+
+@six.add_metaclass(abc.ABCMeta)
+class UpgradeDriver():
+
+    @abc.abstractproperty
+    def supports_upgrade_rollback(self):
+        """Property used to indicate if an implementation supports rolling back
+        an upgrade
+        """
+        pass
+
+    @abc.abstractmethod
+    def pre_host_upgrade_hook(self, host):
+        """Called by the decision engine before a host is upgraded,
+        allows an UpgradeDriver the opportunity to do any operations required
+        before the host is upgraded
+
+        Params
+
+        host: a kostyor.db.models.Host instance, representing the host that is
+        scheduled to be upgraded
+        """
+        pass
+
+    @abc.abstractmethod
+    def pre_service_upgrade_hook(self, service):
+        """Called by the decision engine before a service is upgraded,
+        allows an UpgradeDriver the opportunity to do any operations required
+        before a service is upgraded
+
+        Params
+
+        service: a kostyor.db.models.Service instance, representing the service
+        that is scheduled to be upgraded
+        """
+        pass
+
+    @abc.abstractmethod
+    def post_host_upgrade_hook(self, host):
+        """Called by the decision engine after a host is upgraded,
+        allows an UpgradeDriver the opportunity to do any operations required
+        after the host is upgraded
+
+        Params
+
+        host: a kostyor.db.models.Host instance, representing the host that is
+        scheduled to be upgraded
+        """
+        pass
+
+    @abc.abstractmethod
+    def post_service_upgrade_hook(self, service):
+        """Called by the decision engine after a service is upgraded,
+        allows an UpgradeDriver the opportunity to do any operations required
+        after a service is upgraded
+
+        Params
+
+        service: a kostyor.db.models.Service instance, representing the service
+        that is scheduled to be upgraded
+        """
+        pass
+
+    @abc.abstractmethod
+    def cancel_upgrade(self):
```

Comment:
No objections regarding class methods. The only thing I see, maybe passing an upgrade instance via `__init__` is a bit easy extendible since it doesn't require to change signatures of interface methods. I.e. if tomorrow we will need something except an upgrade instance, we want need to change interface signatures in order to pass that additional param. We only have to change driver instantiation and that's it.

(smth like that)

```python
class MyUpgradeDriver(base.UpgradeDriver):
    def __init__(self, upgrade, another_param):
        self.upgrade = upgrade
        self.another_param = another_param

    def start_upgrade(self):
        # access any needed attribute

driver = MyUpgradeDriver(upgrade_instance, another_param)
driver.start_upgrade()
```